plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.ts

```diff
@@ -508,6 +518,57 @@ export class DefaultProcessingDatabase implements ProcessingDatabase {
     };
   }
 
+  async listAncestors(
+    txOpaque: Transaction,
+    options: ListAncestorsOptions,
+  ): Promise<ListAncestorsResult> {
+    const tx = txOpaque as Knex.Transaction;
+    const { entityRef } = options;
+    const entityRefs = new Array<string>();
+
+    let currentRef = entityRef.toLocaleLowerCase('en-US');
+    for (let depth = 1; depth <= MAX_ANCESTOR_DEPTH; depth += 1) {
+      const rows = await tx<DbRefreshStateReferencesRow>(
+        'refresh_state_references',
+      )
+        .where({ target_entity_ref: currentRef })
+        .select();
+
+      if (rows.length === 0) {
+        if (depth === 1) {
+          throw new NotFoundError(`Entity ${currentRef} not found`);
+        }
+        throw new NotFoundError(
+          `Entity ${entityRef} has a broken parent reference chain at ${currentRef}`,
+        );
+      }
+
+      const parentRef = rows.find(r => r.source_entity_ref)?.source_entity_ref;
```

Comment:
I just realized two things. First: there can be multiple parents. And second: there can be loops. We actually have to deal with these.

So something like

```ts
const parentRefs = rows
  .map(r => r.source_entity_ref)
  .filter((r): r is string => r && !entityRefs.includes(r));
if (!parentRefs.length) { ... }
```

and then we need to keep a list of parents, not a single one, per "layer" traversed upward